### ISSUE TITLE
Take the MPI communicator through an expose call instead of as an initialization parameter

### ISSUE THREAD

**Issue by jbigot** [pdidev/pdi#75]
In GitLab by @jbigot on Mar 13, 2018, 13:34

**Comment by jbigot** [comment]
In GitLab by @ksiero on Jun 5, 2018, 13:36

This approach could be useful for logging. PDI will support different levels of logging for each rank. This is set up through the configuration file. We could add an option to give different levels of logging for communicators or the ranks in specified communicator.

**Comment by jbigot** [comment]
In GitLab by @tpaluszk on Jun 22, 2018, 11:30

Do you mean to move `MPI comm` from `PDI` to `plugins` and then use expose to give the info about communicator to plugins?   
E.g. like this?  
  
code:
```
...
PDI_Init(conf); // init pdi
PDI_expose("FTI_Comm", comm); // expose the comm
do stuff
...
```

config:
```
...
plugins:
  fti_plugin:
    MPI_Comm: FTI_Comm
...
```
And if comm is not exposed or specified in config use `MPI_SELF`/raise an error.  
Or do you have other idea in mind?

**Comment by jbigot** [comment]
@tpaluszk this is exactly the idea I had in mind

**Comment by jbigot** [comment]
In GitLab by @tpaluszk on Jul 9, 2018, 15:34

How would MPI plugin work with other plugins, when they seem to be dependent on MPI plugin (exposing `MPI_Comm`) ?

**Comment by jbigot** [comment]
indeed, this would also lead to dependencies between plugins

**Comment by jbigot** [comment]
* a first step would be to have that without a plugin: add MPI types to PDI core so as to expose communicators like any other data
* in a second step, we can think about adding type plugins

**Comment by jbigot** [comment]
In GitLab by @ksiero on Jul 10, 2018, 09:44

Maybe we could make the MPI_Datatype the same way as MPI_Comm?

code:

```c
...
int array[10];

PDI_Init(conf); // init pdi
MPI_Datatype myComplexType;
MPI_Type_contiguous(10, MPI_INT, &myComplexType);
PDI_expose("myComplexType", &myComplexType, PDI_INOUT); // expose the comm

do_stuff();

PDI_expose("data", array, PDI_OUT);
...
```
config:

```yaml
metadata:
  myComplexType: MPI_Datatype
...
plugins:
  decl_hdf5:
    file: data.h5
    datasets:
      data:
        type: $myComplexType
...
```

Either way user have to define MPI datatype in his code, so maybe we could just use it as metadata. In this case we don't have to write any datatype in config, just save a reference expression and evaluate it when `data` expose is made.

**Comment by jbigot** [comment]
That would be a good option indeed! Hhowever MPI types are not really introspectable IIRC. Also this is unrelated to the `MPI_Comm` question isn't it?

**Comment by jbigot** [comment]
In GitLab by @ksiero on Jul 11, 2018, 09:13

Somehow it is. We can do the `MPI_Comm` and `MPI_Datatype` as one solution. The difference will be only the metadata type.

Do we need MPI types introspection? 
I thinking about creating the new scalar datatypes: `MPI_Comm` and `MPI_Datatype`.
Plugin will use the name (from config) to get descriptor from PDI context.

I am already working on prototype to see if this could work properly :)

**Comment by jbigot** [comment]
In GitLab by @tpaluszk on Jul 12, 2018, 09:24

Wouldn't it also be necessary to split plugins initialization from PDI initialization?  
It's needed e.g. by FTI, because it requires a communicator in init function.

**Comment by jbigot** [comment]
indeed, FTI would need to be initialized later, for example when its communicator is made available

**Comment by jbigot** [comment]
mentioned in commit e51ce8620e99b561e1345c17c2750a5109471df2

**Comment by jbigot** [comment]
mentioned in commit fa9053ab18bdf54c57306bd9de8d087a285ba187

**Comment by jbigot** [comment]
mentioned in commit 1e26de65ba076c2ab44417a2b984a7b48fa929b4

**Comment by jbigot** [comment]
mentioned in commit 1732b220299f67d285a1fb7d24f025c70dd2e61d

**Comment by jbigot** [comment]
mentioned in commit 0d0367b7595e9d08df42731c49777b5a226c1c59

**Comment by jbigot** [comment]
closed via merge request !170

**Comment by jbigot** [comment]
closed via commit 0d0367b7595e9d08df42731c49777b5a226c1c59
